src/main/java/com/monumental/triggers/ModelTrigger.java

```diff
@@ -0,0 +1,135 @@
+package com.monumental.triggers;
+
+import com.monumental.models.Model;
+import org.hibernate.event.spi.*;
+import org.hibernate.persister.entity.EntityPersister;
+import org.springframework.stereotype.Component;
+
+import java.lang.reflect.ParameterizedType;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class wraps the Hibernate event system into an easier-to-use parent class
+ * @param <T> The Model to trigger on - only this Model type will be watched by the subclass
+ */
+@SuppressWarnings("unchecked")
+@Component
+public class ModelTrigger<T extends Model> implements
+        PreInsertEventListener, PostInsertEventListener, PreUpdateEventListener, PostUpdateEventListener,
+        PreDeleteEventListener, PostDeleteEventListener {
+
+    private List<String> propertyNames;
+    private List<Object> state;
+
+    /** Trigger Methods to be Overridden */
+
+    void beforeInsert(T record) {
+
+    }
+    void afterInsert(T record) {
+
+    }
+    void beforeUpdate(T record, T original) {
+
+    }
+    void afterUpdate(T record, T original) {
+
+    }
+    void beforeDelete(T record) {
+
+    }
+    void afterDelete(T record) {
+
+    }
+
+    /** Hibernate Event Listeners */
+
+    @Override
+    public boolean onPreInsert(PreInsertEvent event) {
+        Object entity = event.getEntity();
+        if (entity.getClass().equals(this.getModelClass())) {
+            this.beforeInsert((T) entity);
+        }
+        return false;
```

Comment:
Why return `false`?

Reply:
Returning `true` vetoes the operation (prevents the insert) https://docs.jboss.org/hibernate/orm/4.1/javadocs/org/hibernate/event/spi/PreInsertEventListener.html

I'll add a comment about this